data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java

```diff
@@ -0,0 +1,100 @@
+/*
+ *  Copyright (c) 2023 Fraunhofer Institute for Software and Systems Engineering
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Fraunhofer Institute for Software and Systems Engineering - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.protocol.dsp;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import jakarta.json.Json;
+import jakarta.json.JsonObject;
+import org.eclipse.edc.jsonld.spi.JsonLdKeywords;
+
+import java.util.List;
+import java.util.Objects;
+
+import static org.eclipse.edc.jsonld.spi.JsonLdKeywords.VOCAB;
+import static org.eclipse.edc.protocol.dsp.type.DspPropertyAndTypeNames.DSPACE_PROCESS_ID;
+import static org.eclipse.edc.protocol.dsp.type.DspPropertyAndTypeNames.DSPACE_PROPERTY_CODE;
+import static org.eclipse.edc.protocol.dsp.type.DspPropertyAndTypeNames.DSPACE_PROPERTY_REASON;
+import static org.eclipse.edc.spi.CoreConstants.EDC_NAMESPACE;
+
+/**
+ * A dsp process error.
+ */
+@JsonDeserialize(builder = DspError.Builder.class)
+public class DspError {
+
+    private String type;
+    private String processId;
+    private String code;
+    private List<String> messages;
+
+    public JsonObject toJson() {
```

Comment:
is not a good practice to add serialization details into a value object, we should follow the transformer pattern. Could be a refactor to be addressed in another PR.

Reply:
Agreed, actually we don't need to deserialize this class, it's only for egress. If we need to deserialize errors, we should use a transformer

Reply:
> Agreed, actually we don't need to deserialize this class, it's only for egress. If we need to deserialize errors, we should use a transformer

I guess in the future we would need to deserialize it as well, e.g. to understand the error cause (on the requestor side)

Reply:
When that happens we need to use a transformer and follow the DSP protocol spec. We can't assume an EDC sent the response so there's no need to use the type to deserialize it.

Reply:
`DspError` is following the protocol spec already, isn't it? All the messages we are sending/receiving pass through a specific type and related transformers, I think this should be the same for the errors, our representation doesn't need to be 1:1 with the json message, it just needs to be able to serialize to/deserialize from compliant jsons

Reply:
We shouldn't use DspError to deserialize from the wire. When deserializing an error, I'm also not sure we need a transformer, although we can look into it. The main issue is the DSP error messages are intentionally loosely typed and may contain open-ended error information. This is best represented as Map (or JsonObject), logged in serialized form, and then the code interpreted by the dispatcher. I'm not sure it makes much sense to transform it because it can only be converted into a Map, which it already is.  

I don't think we should use transformers when serializing error messages since this increases complexity significantly for something that can be handled with a simple return type. We would also need to establish what happens when an error transformation fails.